### ISSUE TITLE
swig: OgrePYTHON_wrap.cxx exceeds memory limits during compilation

### ISSUE THREAD

**Issue by simonschmeisser** [OGRECave/ogre#1546]
We run into timeouts and memory starvation during compilation of OgrePYTHON_wrap.cxx on multiple build farms and I'm curious if you have any idea how to split up that file maybe?

Debian buildd for mips64el:
https://buildd.debian.org/status/fetch.php?pkg=ogre-1.12&arch=mips64el&ver=1.12.5%2Bdfsg1-1&stamp=1587297268&raw=0

Debian salsa CI (amd64):
https://salsa.debian.org/games-team/ogre/-/jobs/683313

On mipsel it seems to be a different error:
```
[ 82%] Building CXX object RenderSystems/GL3Plus/CMakeFiles/RenderSystem_GL3Plus.dir/src/OgreGL3PlusPlugin.cpp.o
cd /<<PKGBUILDDIR>>/obj-mipsel-linux-gnu/RenderSystems/GL3Plus && /usr/bin/c++  -DOGRE_GCC_VISIBILITY -DRenderSystem_GL3Plus_EXPORTS -I/<<PKGBUILDDIR>>/RenderSystems/GL3Plus/include -I/<<PKGBUILDDIR>>/RenderSystems/GL3Plus/src/GLSL/include -I/<<PKGBUILDDIR>>/OgreMain/include -I/<<PKGBUILDDIR>>/obj-mipsel-linux-gnu/include -I/<<PKGBUILDDIR>>/RenderSystems/GLSupport/include -I/<<PKGBUILDDIR>>/RenderSystems/GLSupport/include/GLSL  -Wall -Winit-self -Wcast-qual -Wwrite-strings -Wextra -Wundef -Wmissing-declarations -Wno-unused-parameter -Wshadow -Wno-missing-field-initializers -Wno-long-long -g -O2 -fdebug-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -Wformat -Werror=format-security -pipe -Wall -Wdate-time -D_FORTIFY_SOURCE=2 -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -fvisibility=hidden -fvisibility-inlines-hidden   -std=c++11 -o CMakeFiles/RenderSystem_GL3Plus.dir/src/OgreGL3PlusPlugin.cpp.o -c /<<PKGBUILDDIR>>/RenderSystems/GL3Plus/src/OgreGL3PlusPlugin.cpp
virtual memory exhausted: Cannot allocate memory
```
https://buildd.debian.org/status/fetch.php?pkg=ogre-1.12&arch=mipsel&ver=1.12.5%2Bdfsg1-1&stamp=1587263744&raw=0

swig is version 4.0.1.5, ogre is the current release 1.12.5

**Comment by paroj** [comment]
yeah, I also encountered this when building snap packages on the ubuntu builders. My workaround was switching to the clang compiler. I did not figure out how to make swig split one module into multiple compilation units.

**Comment by simonschmeisser** [comment]
Unfortunately even clang has the same problem on MIPSEL architecture which is one of the required architectures for Debian. https://buildd.debian.org/status/package.php?p=ogre-1.12&suite=experimental

I read through the upstream SWIG issue but am not exactly sure I understood how to do the splitting while preserving the current Python api. Did you? I can try some experimenting but if you have clues/a plan that might save some time. Thanks!

**Comment by paroj** [comment]
> how to do the splitting while preserving the current Python api.

I dont think this is possible - i.e. we are hitting a limitation of SWIG here. As a workaround one would have to split OgreMain into sevaral (python) modules. But we currently dont know how to split things: 
- evently? 
- Are there parts that contribute the majority of python wrapper code?

also, other languages (C#, Java) suffer far less from this issue due to less C++ glue code generated.

Ideally this would be solved on the compiler level by this (?):
https://gcc.gnu.org/pipermail/gcc/2020-July/233283.html

**Comment by paroj** [comment]
does  #1877 help?
